core/src/main/scala/org/broadinstitute/dsde/rawls/monitor/BillingAccountChangeSynchronizer.scala

```diff
@@ -0,0 +1,312 @@
+   * sync the billing projects and all workspaces again immediately. If we sync one at a time
+   * we'll have a better chance of doing this once.
+   */
+  def updateBillingAccounts: IO[Unit] =
+    readABillingProjectChange.flatMap(_.traverse_ {
+      syncBillingAccountChange[Kleisli[IO, BillingAccountChange, *]]
+        .handleErrorWith(failChange[Kleisli[IO, BillingAccountChange, *]])
+        .run
+    })
+
+
+  def readABillingProjectChange: IO[Option[BillingAccountChange]] =
+    inTransaction {
+      BillingAccountChanges
+        .latestChanges
+        .unsynced
+        .take(1)
+        .result
+    }.map(_.headOption)
+
+
+  private def syncBillingAccountChange[F[_]](implicit R: Ask[F, BillingAccountChange], M: MonadThrow[F], L: LiftIO[F])
+  : F[Unit] =
+    for {
+      _ <- info("Updating Billing Account on Billing Project in Google")
+      billingProject <- loadBillingProject
+      // the billing probe can only access billing accounts that are defined
+      billingProbeHasAccess <- billingProject.billingAccount match {
+        case Some(accountName) => L.liftIO(gcsDAO.testDMBillingAccountAccess(accountName).io)
+        case None => M.pure(false)
+      }
+
+      updateBillingProjectOutcome <-
+        setBillingProjectBillingAccountInGoogle(billingProject)
+          .attempt
+          .map(Outcome.fromEither)
+
+      _ <- writeUpdateBillingProjectOutcome(billingProbeHasAccess, updateBillingProjectOutcome)
+      updateWorkspacesOutcome <- setWorkspacesBillingAccountInGoogle(billingProject, updateBillingProjectOutcome)
+      _ <- setBillingAccountChangeOutcome(updateBillingProjectOutcome |+| updateWorkspacesOutcome)
+    } yield ()
+
+
+  private def loadBillingProject[F[_]](implicit R: Ask[F, BillingAccountChange], M: Monad[F], L: LiftIO[F])
+  : F[RawlsBillingProject] =
+    for {
+      projectName <- R.reader(_.billingProjectName)
+      projectOpt <- inTransaction(rawlsBillingProjectQuery.load(projectName))
+    } yield projectOpt.getOrElse(throw new IllegalStateException(s"No such billing account $projectName"))
+
+
+  private def setBillingProjectBillingAccountInGoogle[F[_]](billingProject: RawlsBillingProject)
+                                                           (implicit R: Ask[F, BillingAccountChange], M: Monad[F], L: LiftIO[F])
+  : F[Unit] =
+    for {
+      // v1 billing projects are backed by google projects but we have no quick way of telling if
+      // an arbitrary billing project is v1 or not. We don't want to blindly set the billing account
+      // on any google project with an id equal to the billing project name though.
+      // We can use Sam's resource Admin APIs to list the child resources of the billing project. If
+      // a google project resource is listed then we can be sure that this is a v1 billing project.
+      // What follows assumes that the billing project exists and was created successfully.
+      isV1BillingProject <- L.liftIO {
```

Comment:
Indeed it could but I wasn't sure what else would benefit from this (other than our past-lives CI scripts). Would you prefer that I do this? Where should it live?

Reply:
Sure thing, I'll move it into it's own definition :)